+nosnoc/+discrete_time_problem/Cls.m

```diff
@@ -0,0 +1,1600 @@
+              case ConstraintRelaxationMode.ELL_INF % l_inf
+                obj.w.s_terminal_ell_inf = {{'s_terminal_ell_inf', 1}, 0, inf, 1e3};
+
+                g_terminal = [g_terminal-model.lbg_terminal-obj.w.s_terminal_ell_inf();
+                    -(g_terminal-model.ubg_terminal)-obj.w.s_terminal_ell_inf()];
+                obj.g.terminal = {g_terminal, -inf, 0};
+                obj.f = obj.f + obj.p.rho_terminal_p()*obj.w.s_terminal_ell_inf();
+            end
+        end
+
+        % TODO(@anton) is there a way to make this less horribly ugly?
+        function generate_complementarity_constraints(obj)
+            import casadi.*
+            opts = obj.opts;
+            dcs = obj.dcs;
+            model = obj.model;
+            % Do Cross-Complementarity
+            rbp = ~opts.right_boundary_point_explicit;
+
+            % Impulse complementarities
+            if opts.no_initial_impacts
+                start_fe = 2;
+            else
+                start_fe = 1;
+            end
+            for ii=1:opts.N_stages
+                for jj=start_fe:opts.N_finite_elements(ii)
+                    Gij = {};
+                    Hij = {};
+
+                    Gij = [Gij,{obj.w.Lambda_normal(ii,jj)}];
+                    Hij = [Hij,{obj.w.Y_gap(ii,jj) + obj.w.P_vn(ii,jj) + obj.w.N_vn(ii,jj)}];
+                    Gij = [Gij,{obj.w.P_vn(ii,jj)}];
+                    Hij = [Hij,{obj.w.N_vn(ii,jj)}];
+                    if model.friction_exists
+                        switch opts.friction_model
+                          case 'Polyhedral'
+                            Gij = [Gij,{obj.w.Delta_d(ii,jj)}];
+                            Hij = [Hij,{obj.w.Lambda_tangent(ii,jj)}];
+                            Gij = [Gij,{obj.w.Gamma_d(ii,jj)}];
+                            Hij = [Hij,{obj.w.Beta_d(ii,jj)}];
+                          case 'Conic'
+                            Gij = [Gij,{obj.w.Gamma(ii,jj)}];
+                            Hij = [Hij,{obj.w.Beta(ii,jj)}];
+                            switch opts.conic_model_switch_handling
+                              case 'Plain'
+                              case 'Abs'
+                                Gij = [Gij,{obj.w.P_vt(ii,jj)}];
+                                Hij = [Hij,{obj.w.N_vt(ii,jj)}];
+                              case 'Lp'
+                                Gij = [Gij,{obj.w.Alpha_vt(ii,jj)}];
+                                Hij = [Hij,{obj.w.P_vt(ii,jj)}];
+                                Gij = [Gij,{1-obj.w.Alpha_vt(ii,jj)}];
+                                Hij = [Hij,{obj.w.N_vt(ii,jj)}];
+                            end
+                        end
+                    end
+                    obj.G.impulse_comp(ii,jj) = {vertcat(Gij{:})};
+                    obj.H.impulse_comp(ii,jj) = {vertcat(Hij{:})};
+                end
+            end
+            % contact complementarities
+            if opts.use_fesd
```

Comment:
we may not even need this if, as false is meaningless and caught earlier.

Reply:
If we are dropping non-fesd support for Cls we can probably catch that at the `ocp.Solver` stage.